model/model.go

```diff
@@ -272,18 +215,27 @@ func canNil(t reflect.Type) bool {
 		t.Kind() == reflect.Slice
 }
 
-func Forward(m Model, optsFuncs ...OptionsFunc) (ml.Tensor, error) {
-	var opts Options
-	for _, optsFunc := range optsFuncs {
-		optsFunc(m, &opts)
+func Forward(ctx ml.Context, m Model, opts Options) (ml.Tensor, error) {
+	if len(opts.Positions) != len(opts.Sequences) {
+		return nil, fmt.Errorf("length of positions (%v) must match length of seqs (%v)", len(opts.Positions), len(opts.Sequences))
+	}
+
+	if len(opts.Positions) < 1 {
+		return nil, errors.New("batch size cannot be less than 1")
+	}
+
+	cache := m.Config().Cache
+	if cache != nil {
+		err := cache.StartForward(ctx, opts.Positions, opts.Sequences)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	ctx := m.Backend().NewContext()
 	t, err := m.Forward(ctx, opts)
 	if err != nil {
 		return nil, err
 	}
-	defer ctx.Close()
```

Comment:
does `cache.Close()` need to be called explicitly in the runner?

Reply:
Not sure if you are referring to `cache` or `ctx` based on the quoted code.

`ctx` does need to be closed in the runner rather than here because otherwise the returned tensor is no longer valid after close.

`cache` has a `Close` function but we never really deallocate the model until the entire runner killed, at which point everything will be freed. I'm probably going to move the resource allocation for caches to the runner so that they can share memory better, which will also make it easier to cleanup everything at the same time. However, we still realistically won't free the memory until process exit.